tests/utils/mod.rs

```diff
@@ -0,0 +1,184 @@
+        cid: Cid,
+        respond_to: oneshot::Sender<Result<Vec<u8>, Error>>,
+    },
+}
+
+struct TestBitswapWorker {
+    swarm: BitswapSwarm,
+    cmd: mpsc::UnboundedReceiver<NodeCommand>,
+    queried_cids: FnvHashMap<QueryId, oneshot::Sender<Result<Vec<u8>, Error>>>,
+    dials_requested: FnvHashMap<PeerId, oneshot::Sender<Result<(), DialError>>>,
+}
+
+impl TestBitswapWorker {
+    async fn run(&mut self) {
+        loop {
+            select! {
+                ev = self.swarm.select_next_some() => {
+                    self.on_event(ev);
+                },
+                cmd = self.cmd.recv() => if let Some(cmd) = cmd {
+                    self.on_cmd(cmd).await;
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn on_event(&mut self, ev: SwarmEvent<Event>) {
+        match ev {
+            SwarmEvent::Behaviour(bev) => match bev {
+                Event::GetQueryResponse { query_id, data } => {
+                    let tx = self.queried_cids.remove(&query_id).unwrap();
+                    tx.send(Ok(data)).unwrap();
+                }
+                Event::GetQueryError { query_id, error } => {
+                    let tx = self.queried_cids.remove(&query_id).unwrap();
+                    tx.send(Err(error)).unwrap();
+                }
+            },
+            SwarmEvent::ConnectionEstablished { peer_id, .. } => {
+                if let Some(tx) = self.dials_requested.remove(&peer_id) {
+                    // receiver might have been dropped already if caller doesn't care about
+                    // waiting for connecion to be established
+                    let _ = tx.send(Ok(()));
+                }
+            }
+            _ => (),
+        }
+    }
+
+    async fn on_cmd(&mut self, cmd: NodeCommand) {
+        match cmd {
+            NodeCommand::Connect {
+                addr,
+                peer_id,
+                notify_connected,
+            } => {
+                if let e @ Err(_) = self.swarm.dial(addr.clone()) {
+                    notify_connected.send(e).unwrap();
+                } else {
+                    self.dials_requested.insert(peer_id, notify_connected);
+                }
+            }
+            NodeCommand::WaitForCid { cid, respond_to } => {
+                let query_id = self.swarm.behaviour_mut().get(&cid);
+                self.queried_cids.insert(query_id, respond_to);
+            }
+        }
+    }
+}
+
+pub async fn spawn_node(store: Option<InMemoryBlockstore<CID_SIZE>>) -> TestBitswapNode {
+    let store = store.unwrap_or_default();
+
+    let mut swarm = SwarmBuilder::with_new_identity()
+        .with_tokio()
+        .with_tcp(
+            tcp::Config::default(),
+            libp2p_noise::Config::new,
+            libp2p_yamux::Config::default,
+        )
+        .unwrap()
+        .with_behaviour(|_key| beetswap::Behaviour::<CID_SIZE, _>::new(store))
+        .unwrap()
+        .build();
```

Comment:
We can also use [`libp2p_swarm_test::SwarmExt`](https://docs.rs/libp2p-swarm-test/latest/libp2p_swarm_test/trait.SwarmExt.html) instead of creating a TCP connection, but lets do it later on.